tests/pyfunc/test_pyfunc_schema_enforcement.py

```diff
@@ -914,3 +914,98 @@ def test_schema_enforcement_for_list_inputs():
+        "b": 1,
+        "c": True,
+        "d": 1.0,
+        "e": [1, 2, 3],
+        "f": (1, 2, 3),
+        "g": b"byte_g",
+        "h": {"test_k": "test_v"},
+    }
+    test_schema = ParamSchema(
+        [
+            ParamSpec("a", "str", ""),
+            ParamSpec("b", "int", 1),
+            ParamSpec("c", "bool", False),
+            ParamSpec("d", "float", 1.0),
+            ParamSpec("e", "list", []),
+            ParamSpec("f", "tuple", (1,)),
+            ParamSpec("g", "bytes", b""),
+            ParamSpec("h", "dict", {}),
+        ]
+    )
+    assert _enforce_parameters_schema(test_parameters, test_schema) == test_parameters
+
+    # Add default values if the parameter is optional but not provided
+    test_parameters = {"a": "str_a"}
+    test_schema = ParamSchema([ParamSpec("a", "str"), ParamSpec("b", "int", 1)])
+    updated_parameters = {"b": 1}
+    updated_parameters.update(test_parameters)
+    assert _enforce_parameters_schema(test_parameters, test_schema) == updated_parameters
+
+    # Converting parameters keys to string if it is not
+    test_parameters = {1: 1.0}
+    test_schema = ParamSchema([ParamSpec("1", "float")])
+    assert _enforce_parameters_schema(test_parameters, test_schema) == {"1": 1.0}
+
+    # Raise Exception if non-optional parameters has default value
+    with pytest.raises(
+        MlflowException, match=r"Default value has to be None for required ParamSpec 'a'"
+    ):
+        ParamSpec("a", "str", "default_value", False)
+
+    # Raise Exception if non-optional parameters is not provided
+    test_parameters = {}
+    test_schema = ParamSchema(
+        [
+            ParamSpec("a", "str", optional=False),
+        ]
+    )
+    with pytest.raises(MlflowException, match=r"Missing required parameter: a"):
+        _enforce_parameters_schema(test_parameters, test_schema)
+
+    # Raise error if parameters is not dictionary
+    with pytest.raises(MlflowException, match=r"Parameters must be a dictionary. Got type 'int'."):
+        _enforce_parameters_schema(100, test_schema)
+
+    # Raise error if default value is not consistent with provided type
+    with pytest.raises(MlflowException, match=r"Invalid default value for ParamSpec"):
+        test_schema = ParamSchema([ParamSpec("a", "str", default=1.0)])
+
+    # Raise error if invalid parameters are passed
+    test_parameters = {"a": 100, "b": (1, 2), "c": b"test"}
+    test_schema = ParamSchema(
+        [
+            ParamSpec("a", "int"),
+            ParamSpec("b", "list"),
+            ParamSpec("c", "str"),
+        ]
+    )
+    with pytest.raises(  # pylint: disable=[pytest-raises-multiple-statements, pytest-raises-without-match]
+        MlflowException
+    ) as e:
+        _enforce_parameters_schema(test_parameters, test_schema)
+        assert e.match(r"('b', 'Invalid type for parameter b: expected list but got tuple')")
+        assert e.match(r"('c', 'Invalid type for parameter c: expected str but got bytes')")
```

Comment:
/ai I'm a python beginner. Can you tell me why `pytest.raises` should NOT contain multiple statements.

Reply:
A perfect explanation.